### PR TITLE
[#1732] Use timeout to wait for disposition update

### LINKED ISSUE #1732: AMQP Adapter: Use timeout when sending command to device
When the AMQP adapter sends a command to a device, the adapter waits for the delivery update sent from the device and then itself sends a delivery update to the northbound application.

There is currently no timeout being used when waiting for the delivery update sent from the device.
This still needs to be implemented.

### PR DESCRIPTION
**Pull request by Alfusainey**
AMQP adapter now uses timeout to wait for a disposition update
in response to a command sent to a device.

fixes #1732

TODO: document in the amqp  adapter user guide

Signed-off-by: Alfusainey Jallow <alf.jallow@gmail.com>

### CODE REVIEW

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java [hunk 1]
```diff
@@ -892,7 +892,20 @@ protected void onCommandReceived(final TenantObject tenantObject, final ProtonSe
...
+            // timeout expired -> release command
+            commandContext.getCurrentSpan().log("timeout waiting for delivery update from device");
+            commandContext.release();
+        });
+
         sender.send(msg, delivery -> {
+            // disposition received -> cancel timer
+            vertx.cancelTimer(timerId);
+
             // release the command message when the device either
```
Comment (calohmn) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java:909]: A check is needed here whether the timeout has already occurred, skipping the delivery update handling below in that case.

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java [hunk 2]
```diff
@@ -892,7 +892,20 @@ protected void onCommandReceived(final TenantObject tenantObject, final ProtonSe
         }
 
         // TODO time out waiting for disposition update
+        final Long timerId = vertx.setTimer(getConfig().getDeliveryUpdateTimeout(), tid -> {
```
[the remark this answers is not in the document]
Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java:895]: > The case that the deliveryUpdateTimeout is 0 should be taken into account here

how about only allowing values `> 0` for this config property?

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterProperties.java [hunk 3]
```diff
@@ -137,4 +143,30 @@ public final void setIdleTimeout(final int timeout) {
...
+    }
+
+    /**
+     * Sets the time to wait for a delivery update from a device before the AMQP sender link is closed.
+     * 
+     * @param deliveryUpdateTimeout The timeout value in milliseconds.
+     * 
+     * @throws IllegalArgumentException if the timeout value is negative.
+     */
+    public final void setDeliveryUpdateTimeout(final long deliveryUpdateTimeout) {
```
[the remark this answers is not in the document]
Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterProperties.java:166]: make sense to me 👍

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (calohmn) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java:895]: I would rather keep it in line with the handling of the `sendMessageTimeout` value (and e.g. connection `requestTimeout`) and allow `0` for an infinite timeout.

**Comment by Alfusainey** [comment]
@calohmn i made the requested changes. can you please take a look?

File: site/documentation/content/admin-guide/amqp-adapter-config.md [hunk 4]
```diff
@@ -27,6 +27,7 @@ The following table provides an overview of the configuration variables and corr
...
 | `HONO_AMQP_DEFAULTS_ENABLED`<br>`--hono.amqp.defaultsEnabled` | no | `true` | If set to `true` the protocol adapter uses *default values* registered for a device to augment messages published by the device with missing information like a content type. In particular, the protocol adapter adds default values registered for the device as (application) properties with the same name to the AMQP 1.0 messages it sends downstream to the AMQP Messaging Network. |
 | `HONO_AMQP_IDLE_TIMEOUT`<br>`--hono.amqp.idleTimeout` | no | `60000` | The time interval (milliseconds) to wait for incoming traffic from a device before the connection should be considered stale and thus be closed. Setting this property to `0` prevents the adapter from detecting and closing stale connections. |
+| `HONO_AMQP_SEND_MESSAGE_TO_DEVICE_TIMEOUT`<br>`--hono.amqp.sendMessageToDeviceTimeout` | no | `1000` | The time interval (milliseconds) to wait for a device to acknowledge receiving a (command) message before the AMQP link used for sending the message will be close. Setting this property to `0` means the adapter waits indefinitely for a device to  acknowledge receiving the message. |
```
Comment (calohmn) [inline site/documentation/content/admin-guide/amqp-adapter-config.md:30]: "will be closed"

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterProperties.java [hunk 5]
```diff
@@ -137,4 +143,30 @@ public final void setIdleTimeout(final int timeout) {
     public final int getIdleTimeout() {
         return this.idleTimeout;
     }
+
+    /**
+     * Gets the time to wait for a delivery update from a device before the AMQP sender link to the
+     * device is close.
```
Comment (calohmn) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterProperties.java:149]: "is closed"

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterProperties.java [hunk 6]
```diff
@@ -137,4 +143,30 @@ public final void setIdleTimeout(final int timeout) {
...
+     * <p>
+     * The default value of this property is {@link #sendMessageToDeviceTimeout}.
```
Comment (calohmn) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterProperties.java:151]: The link should probably go to `DEFAULT_SEND_MESSAGE_TO_DEVICE_TIMEOUT`.

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java [hunk 7]
```diff
@@ -891,42 +892,68 @@ protected void onCommandReceived(final TenantObject tenantObject, final ProtonSe
...
+                final Exception ex = new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE,
+                        "timeout waiting for delivery update");
+                closeLinkWithError(sender, ex, commandContext.getCurrentSpan());
+                // timeout expired -> release command
+                commandContext.getCurrentSpan().log("timeout waiting for delivery update from device");
+                commandContext.release();
+            } else {
+                log.trace("command is already settled and downstream application notified");
+            }
+        });
```
Comment (calohmn) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java:910]: Can you run the code-formatter here? I think the code will be better readable then.

**Comment by sophokles73** [comment]
@calohmn are you going to merge this PR? Don't forget to squash ...

**Pull request merged**